src/software/jetson_nano/services/network/network.cpp

```diff
@@ -74,3 +77,66 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
                      << "%";
     }
 }
+
+void NetworkService::logNewPrimitiveSet(TbotsProto::PrimitiveSet input)
+{
+    /* TODO: THIS TRACKS THE RECEIVED PRIMITIVES
+     *  - make sure it is in the lock
+     *  - log the new PRIMITIVE SET PROTO within a deque if it is valid (see primitive_tracker.send())
+     *  - update the timestamp on each primitive set to the current epoch time
+     */
+
+    if (primitive_set_tuple_rtt.size() >= 50)
+    {
+        LOG(WARNING) << "Too many primitive sets logged for round-trip calculations, halting log process";
+        return;
+    }
+
+    if (!primitive_set_tuple_rtt.empty()
+            && input.sequence_number() <= std::get<0>(primitive_set_tuple_rtt.back()))
+    {
+        // If the proto is older than the last received proto, then ignore it
+        return;
+    }
+    primitive_set_tuple_rtt.emplace_back(
+            std::tuple(input.sequence_number(),
+                       getCurrentEpochTimeInSeconds(),
+                       input.time_sent().epoch_timestamp_seconds())
+                       );
+}
+
+void NetworkService::updatePrimitiveSetLog(TbotsProto::RobotStatus &robot_status)
+{
+    /**
+     * TODO: Send return RobotStatus msg to Thunderscope
+     *      1) Traverse deque containing all received primitive_sets
+     *      2) If the last_handled_primitive_set for current robot_status is stored in the deque:
+     *         - delete all earlier sets
+     *         - stop & calculate the processing counter
+     *         - store the new omit_thunderloop_processing_time_sent = time_sent + processing duration
+     */
+
+    uint64_t seq_num = robot_status.last_handled_primitive_set();
+    while (seq_num <= std::get<0>(primitive_set_tuple_rtt.back()))
+    {
+        if (std::get<0>(primitive_set_tuple_rtt.front()) == seq_num)
+        {
+            double received_epoch_time_seconds = std::get<1>(primitive_set_tuple_rtt.front());
+            double processing_time_seconds = getCurrentEpochTimeInSeconds() - received_epoch_time_seconds;
```

Comment:
why not just store `processing_time_seconds` in the proto?

Reply:
Yup, since even if we stored the processing time within `RobotStatus`, we would still need to store the additional field containing the time sent of the original primitive set (since this is different from the time sent of the `RobotStatus`)